### ISSUE TITLE
DHARMa integration

### ISSUE THREAD

**Issue by florianhartig** [jslefche/piecewiseSEM#197]
Hi Jon,

I'm the developer of [DHARMa](https://cran.r-project.org/web/packages/DHARMa/index.html) and incidentally just attending a workshop on SEMs, which made me think about the possibility to integrate DHARMa with piecewiseSEM.

I have looked at the structure of your package, and I see two routes one could go

1. **Integrating DHARMa into piecewiseSEM:** most of the models that are supported by piecewiseSEM could also be checked with DHARMa (except nlme). So, it would be possible to add switch in piecewiseSEM and return DHARMa residuals for each model, or p-values for DHARMa omnibus residual tests that could be superimposed on the DAG or displayed in piecewiseSEM.

2. **Integrating piecewiseSEM into DHARMa:** to support piecewiseSEM, minimum requirements are described here https://github.com/florianhartig/DHARMa/wiki/Adding-new-R-packages-to-DHARMA 

Option 2 would allow to create omnibus for a tested SEM from DHARMa, e.g. global dispersion of spatial autocorrelation tests. Given the nature of the package, however, it seems to me more logical to consider each link individually (because averaging over them might mask problems), which would suggest to me that it would be more logical to go via route 1. I think what could added easily is implement

a) the ability to return DHARMa instead of normal residuals per link
b) the option to impose the results of testResiduals (3 general tests) or testUniformity (tests just for distribution) on the links in the dag, e.g. in the summary() or in the plot() function, to alert users to connections that are problematic

**Comment by florianhartig** [comment]
Hi, I just played a bit around with that - here an example of how one could modify the residuals function in piecewiseSEM to support DHARMa ... maybe one should check in the loop if the fitted models  are  supported by DHARMa, and fall back to standard otherwise. 

```
library(piecewiseSEM)

mod = psem(
  lm(rich ~ distance + elev + abiotic + age + hetero + firesev + cover, data =  keeley),
  lm(firesev ~ elev + age + cover, data =  keeley), 
  lm(cover ~ age + elev + abiotic, data =  keeley)
)

summary(mod)
plot(mod)


library(DHARMa)
residuals2 <- function(fittedPSEM, type = "DHARMa"){
  if(type == "standard"){
    return(residuals(fittedPSEM))
  }
  if(type == "DHARMa"){
    sim = list()
    res = list()
    for(i in 1:(length(fittedPSEM) - 1)){
      sim[[i]] <- simulateResiduals(fittedPSEM[[i]])
      res[[i]] = sim[[i]]$scaledResiduals
    }
    res = matrix(unlist(res),ncol=length(res),byrow=F)
    return(list(residuals = res, DHARMaSimulations = sim))
  }
}

res = residuals2(mod)
res$residuals
# plots and tests for model 1
plot(res$DHARMaSimulations[[1]])
testResiduals(res$DHARMaSimulations[[1]])
# in principle, it would be possible to plot residuals also against a particular path, or test for homogeneity along this path
```

**Comment by jslefche** [comment]
This is clever and timely. I would be interested in integrating more robust tests of assumptions in the package –looping over the list seems the most applicable since there are no formal omnibus tests across multiple models. Its good you opened an issue as that will keep it on my radar

____________________________
Jonathan S. Lefcheck, Ph.D.
Tennenbaum Coordinating Scientist
MarineGEO: https://marinegeo.si.edu/
Smithsonian Institution
Phone: +1 (443) 482-2443
www.jonlefcheck.net

**Comment by jebyrnes** [comment]
Yup - I've just lapplied over objects in the past. This approach should work great.

**Comment by florianhartig** [comment]
I guess you would have to decide if you just want to return the scaled residuals, or a list with residuals and the DHARMa objects (as I did here).

There forme makes it compatible to the normal residuals, but one looses the possibility to run the tests implemented in DHARMa on the returned residuals.
